### ISSUE TITLE
make pkg_tar's "./" mandatory prefix optional

### ISSUE THREAD

**Issue by beasleyr-vmw** [bazelbuild/rules_pkg#50]
### Description of the problem / feature request:

Files embedded in a tar created by `pkg_tar` include a mandatory "./" prefix. I would like to make this prefix optional.

### Feature requests: what underlying problem are you trying to solve with this feature?

This is an interoperability/migration problem. My request incurs tech debt to finance Bazel adoption.

I'm trying to migrate a constellation of builds from Maven to Bazel. When I port project X to Bazel, I still need to support some number of preexisting downstream builds. These builds extract subsets of files from tars using whitelists which are **not** anchored with a leading "./". When fed a tar generated by Bazel's `pkg_tar`, a whitelist of "foo" doesn't match "./foo", and so the untar action is a no-op.

Rather than have to discover/fix every trivial difference downstream, I would rather start by producing Bazel bits which are drop-in replacements for Maven bits. Our internal teams can then coordinate among themselves to pay down the debt and remove use of any interop flags.

For the record, a change to remove the "./" prefix was [committed back in 2017][1] but soon [rolled back][2]. I think an _option_ to remove the prefix would help folks like myself w/o risk of breaking existing Bazel users.

[1]: https://github.com/bazelbuild/bazel/commit/72ab9322ac4ea6b743bc304b7b5d05f8b5b56239#diff-002c180b932b0dbae88c2559af105850
[2]: https://github.com/bazelbuild/bazel/commit/36178dddc45ee39e302516232c83b0fbc4c5dc37#diff-002c180b932b0dbae88c2559af105850

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

n/a

### What operating system are you running Bazel on?

Linux - CentOS 6.6 and 7.2
macOS - High Sierra (I think)

### What's the output of `bazel info release`?

"development version"

### If `bazel info release` returns "development version" or "(@non-git)", tell us how you built Bazel.

We track Bazel's Git repo and apply a small number of patches atop official releases. We're currently building against 0.21.0.

### What's the output of `git remote get-url origin ; git rev-parse master ; git rev-parse HEAD` ?

n/a

###  Have you found anything relevant by searching the web?

Yes -- https://github.com/bazelbuild/bazel/commits/master/tools/build_defs/pkg/archive.py:
- [Don't prepend ./ to filenames in pkg_tar rule](https://github.com/bazelbuild/bazel/commit/72ab9322ac4ea6b743bc304b7b5d05f8b5b56239#diff-002c180b932b0dbae88c2559af105850)
- [Automated rollback of commit 39448ebab889a6b5b1ceabcf932adfec9098dfe7.](https://github.com/bazelbuild/bazel/commit/36178dddc45ee39e302516232c83b0fbc4c5dc37#diff-002c180b932b0dbae88c2559af105850)

I chatted w/ @ixdy briefly in Slack about this [on Fri Jan 25](https://bazelbuild.slack.com/archives/CA31HN1T3/p1548438200380200). (Thanks!)

**Comment by kylecordes** [comment]
Now that 1.0 is here certainly anyone can agree that this should be implemented in a nonbreaking way.

But it might also be a candidate to change the default for a future major version. Thinking of tar files across numerous contexts for many years, my impression is that mandatory prefix with “.” is relatively unusual. Not prefixing (by default) would match the expectations of more adopters.

**Comment by ixdy** [comment]
@kylecordes rules_pkg is decoupled from bazel, so it probably isn't strictly necessary to follow the same deprecation strategy.

**Comment by gebn** [comment]
I'm not sure if it's a related issue, but forcing `./` seems to result in trying to modify attributes of the present working directory, which isn't always possible. For example, if I try to extract to `/tmp`, I get:

```
$ tar tvf archive.tar.gz 
drwxr-xr-x 65534/65534       0 2000-01-01 00:00 ./
drwxr-xr-x 65534/65534       0 2000-01-01 00:00 ./bmc_exporter/
-r-xr-xr-x 65534/65534 18340890 2000-01-01 00:00 ./bmc_exporter/bmc_exporter
$ tar xf archive.tar.gz                                                                                                                                     
tar: .: Cannot utime: Operation not permitted                                                                                                                            
tar: .: Cannot change mode to rwxr-xr-t: Operation not permitted                                                                                                         
tar: Exiting with failure status due to previous errors
```

If archives did not have to contain `./`, presumably this would be avoided.

**Comment by lalten** [comment]
Any progress on this? would be great to have a `prepend_dotslash` attr.

**Comment by aiuto** [comment]
I have not had a chance to work on it. 

I agree with @kylecordes that the initial "./" is strange, and also with @beasleyr-vmw that if we change the default that will break some people, and his analysis about using the package_root.

How does everyone feel about this new behavior.
- "./" will not be in the path by default.
- pkg_tar and pkg_zip will both add a `package_base` option which
   - prepends package_base to all paths in the archive (adding / if needed)
   - adds package_base to the archive
      - this parallels behavior of command line tools when you say 'zip -r out.zip foo'
      - some day we might follow that with the ability to specify owner and perms
   - the name package_base is subject to debate because is should be the same word as used in (the upcoming) pkg_filegroup to rebase a set of files.
 
If you are broken by the change in default, you can fix that by adding a package_base. Yes, this is a breaking change but we are still at 0.x, and we know the next is 0.3.0 rather than 0.2.6.

Would this be acceptable?
Does someone want to prepare a PR?

**Comment by lalten** [comment]
Sounds good!

For the record, here's my workaround using a genrule. Using pkg_tar would definitely be more elegant!
```
genrule(
    name = "packaged_target",
    srcs = [":mytarget"],
    outs = ["archive.tar.gz"],
    cmd = " && ".join([
        "pushd $(location :mytarget)",
        "export GZIP=--no-name",  # do not save time stamp in order to stabilize archive hashes
        "tar cfz tmp.tar.gz *",
        "popd",
        "mv $(location :mytarget)/tmp.tar.gz $@",
    ]),
)
```

**Comment by mkaufmann** [comment]
I was happy to see the fix for this released today but it doesn't seem to have the right semantics.

I need to package a file so that is shows the path as 
> Charts.yaml
otherwise the tar won't worm with the Helm tool. 

I thought I would get this behavior when using
> package_base = ""
but that results in a tar with a "/" prefix
> /Charts.yaml

Given that this is released now I sadly don't know how to fix this in a backwards compatible way

**Comment by aiuto** [comment]
Yup. This is not fixed right.  The behavior for this

```
pkg_tar(
    name = "test_no_package_base",
    srcs = [
        "index.html",
        "images/foo.jpg",
    ],
)
```
should be 
```
$ tar tvf ../bazel-bin/test_no_package_base.tar
-r-xr-xr-x  0 0      0           7 Dec 31  1999 index.html
-r-xr-xr-x  0 0      0           6 Dec 31  1999 images/foo.jpg
```
but is
```
tar tvf ../bazel-bin/test_no_package_base.tar
drwxr-xr-x  0 0      0           0 Dec 31  1999 ./
-r-xr-xr-x  0 0      0           7 Dec 31  1999 ./index.html
-r-xr-xr-x  0 0      0           6 Dec 31  1999 ./foo.jpg
```

So there are 2 things wrong.  The images folder is being flattened away.

**Comment by nacl** [comment]
So, how does this fit in with the `pkg_filegroup` framework?  To me, it seems like we should go with the breaking behavior: `./` prefix should need to be explicitly provided to `pkg_files` and/or `pkg_filegroup` for it to appear, with the understanding that some packaging rules may ignore it.

Following from this, does this mean we should change the behavior of `package_base`?  Also another breaking change I'd be happy to look at when I get some time at the end of the month.

**Comment by nacl** [comment]
Hacked on this a little bit.  Some observations:

- `/`-anchored tar files and `archive.py` do not get along well.  Fortunately, they don't seem to be particularly well-supported, or needed.  I'll see if my use case requires them, but we might just be able to get away with not supporting it at all.
- In fact, the whole tar archive writer in `archive.py` probably should be rewritten.  A particularly nice solution would be to have something that schedules the creation of all paths prior to actually assembling the tarfile.  This would also have the nice property of being extensible to things beyond just tarfiles.  It might require, among other things , #385.
- The `package_base` problem was relatively easy to fix and test.  I can have a PR made for this soonish.

The question still remains: do we do with `pkg_filegroup`s and this?  This is a path manipulation, so, strictly speaking, paths should be defined in the incoming `pkg_filegroup`s.  This then would require users to do something like:

```starlark
pkg_filegroup(
    name = "reroot_tar_inputs",
    srcs = [":all_the_srcs"],
    prefix = "./",    
)
```

Or prefix everything with `./` themselves in `pkg_files` and friends.  I don't see a big problem requiring this -- it keeps the path interface clean.

Another option would be to put this rooting stuff back in `pkg_tar` even with `pkg_filegroup`s.  It would be convenient at the price of cluttering the interface a bit.  I still vote for keeping everything in `pkg_files` and friends due to the minimal boilerplate cost.  Users can easily macro their way around it if needed.

Remaining questions:

- Do we want to support rooted tarfiles, or any rooted package?
- What do we do about the `pkg_filegroup` problem, above?

**Comment by aiuto** [comment]
I'm working on this now.

The least surprising solution that I am going for is
-  "./" is not prepended to paths.
- pkg_tar can add a directory prefix, which will get prepended.  

This way, you can have pkg_* rules do whatever they want to  make a path, or you can use legacy constructions and add a path at file writing time.

There is some layered stupidity in archive.py, but I am note ready to remove it yet. I think the debian packaging needs it.  My patfh forward is to remove the mistake of insisting on a leading './' in archive in one PR, then remove the identical mistake at the pkg_tar level.

**Comment by aiuto** [comment]
May be fixed by #554, except I'm not putting much work into the ability to set the package prefix to "./". 
I have yet to see a good reason why that is ever a desired behavior - other than backwards compatibility to broken behavior.
